### PR TITLE
feat(calc): implement DefenceCalculator for defensive stat calculations

### LINKED ISSUE #24: Implement DefenceCalculator (CalcDefence)
## Description
Implement defensive stat calculations (EHP, mitigation, resistances).

## Reference
Original PoB: `src/Modules/CalcDefence.lua` (233 KB)

## Tasks
- [ ] Create `src/engine/calculator/DefenceCalculator.ts`
- [ ] Calculate Life, ES, Mana pools
- [ ] Calculate resistances (fire, cold, lightning, chaos)
- [ ] Calculate armour and physical reduction
- [ ] Calculate evasion and dodge
- [ ] Calculate block chance
- [ ] Calculate EHP (Effective HP)
- [ ] Handle Mind over Matter
- [ ] Calculate damage taken modifiers

## Defence Layers
1. Avoidance: Evasion, Dodge, Block
2. Mitigation: Armour, Resistances, Reduction
3. Pool: Life, ES, Mana

## EHP Calculation
EHP considers all defence layers and their interaction.

## Acceptance Criteria
- Matches PoB calculations
- All defence types calculated
- EHP breakdown available
- Performance < 30ms

## Files to Create
- `src/engine/calculator/DefenceCalculator.ts`
- `src/engine/calculator/MitigationCalculator.ts`

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Implements comprehensive defensive stat calculations (CalcDefence) for the calculation engine.

### New Files
- `src/engine/calculator/DefenceCalculator.ts` - Main calculator with all defensive stat calculations
- `test/unit/engine/calculator/DefenceCalculator.test.ts` - 55 unit tests covering all calculations

### Modified Files
- `src/engine/calculator/index.ts` - Export new calculator functions and types

#### Resource Pools
- Life calculation with level scaling (38 + 12/level) and Str bonus (+0.5 life/Str)
- Energy Shield with Int bonus (+0.2% ES/Int)
- Mana calculation with level scaling (34 + 6/level) and Int bonus (+0.5 mana/Int)

#### Resistances
- Fire, Cold, Lightning, Chaos resistances
- Default cap at 75%, can be raised up to 90% maximum
- Overcap tracking for resistance optimization

#### Physical Mitigation
- Armour rating with INC/MORE modifiers
- Physical DR formula: `Armour / (Armour + 5 * Damage)`, capped at 90%
- Returns function for hit-dependent calculations

#### Avoidance
- Evasion rating with Dex bonus (+0.2% Evasion/Dex)
- Evade chance formula vs enemy accuracy
- Attack block and spell block with 75% cap
- Spell suppression (PoE2) with effect scaling

#### Damage Taken Modifiers
- Reduced damage taken (additive)
- Less damage taken (multiplicative)

#### Mind over Matter
- Mana contribution to effective pool
- Proper ratio calculation for EHP impact

#### Effective HP (EHP)
- Per-damage-type EHP (Physical, Fire, Cold, Lightning, Chaos)
- Elemental EHP average
- Accounts for all mitigation layers

## Testing

- [x] 55 unit tests added covering:
  - Pool calculations with modifiers
  - Resistance capping and overcap
  - Armour DR formula accuracy
  - Evasion chance calculations
  - Block and suppression caps
  - EHP calculations with various defenses
  - Edge cases (empty mods, level 1, high values)
- [x] All tests passing
- [x] Build succeeds
- [x] Lint passes (no errors in new files)

## Interface

```typescript
interface DefenceResult {
  // Pools
  life: number;
  lifeBase: number;
  energyShield: number;
  mana: number;
  
  // Resistances
  fireRes: ResistanceResult;
  coldRes: ResistanceResult;
  lightningRes: ResistanceResult;
  chaosRes: ResistanceResult;
  
  // Physical mitigation
  armour: number;
  physicalDR: (hitDamage: number) => number;
  physicalDRDisplay: number;
  
  // Avoidance
  evasion: number;
  evadeChance: number;
  blockChance: number;
  spellBlockChance: number;
  spellSuppressionChance: number;
  spellSuppressionEffect: number;
  
  // Damage taken modifiers
  lessDamageTaken: number;
  reducedDamageTaken: number;
  
  // Mind over Matter
  hasMindOverMatter: boolean;
  mindOverMatterPercent: number;
  
  // EHP
  ehp: number;
  ehpPhysical: number;
  ehpElemental: number;
  ehpChaos: number;
  ehpFire: number;
  ehpCold: number;
  ehpLightning: number;
}
```

Closes #24

### CODE REVIEW

File: src/engine/calculator/DefenceCalculator.ts [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+  const maxCapBonus = Math.floor(maxCapResult.value);
+
+  // Calculate actual cap (default + bonus, clamped to absolute max)
+  const maxCap = Math.min(
+    RESISTANCE_CONSTANTS.DEFAULT_RES_CAP + maxCapBonus,
+    RESISTANCE_CONSTANTS.MAX_RES_CAP
+  );
+
+  // Clamp to cap
+  const value = Math.min(Math.max(uncapped, 0), maxCap);
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 953ef16. Removed the Math.max(uncapped, 0) to allow negative resistance values.

File: src/engine/calculator/DefenceCalculator.ts [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+
+/** Base accuracy formula: Base + (Level * 2) */
+const ENEMY_BASE_ACCURACY = 200;
+const ENEMY_ACCURACY_PER_LEVEL = 2;
+
+/** Reference hit damage for EHP and display calculations */
+const REFERENCE_HIT_DAMAGE = 5000;
+
+/** Maximum physical damage reduction from armour */
+const MAX_ARMOUR_DR = 0.9;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 953ef16. All magic numbers now use DEFENCE_CONSTANTS prefix consistently.

File: src/engine/calculator/DefenceCalculator.ts [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+
+// ============================================================================
+// Types
+// ============================================================================
+
+/**
+ * Resistance result with capped value and overcap tracking.
+ */
+export interface ResistanceResult {
+  /** Capped resistance value (0-max) */
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 953ef16. Updated documentation to say 'can be negative, capped at max'.

File: test/unit/engine/calculator/DefenceCalculator.test.ts [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+      expect(result.mindOverMatterPercent).toBe(0);
+    });
+
+    it('should cap MoM at 100%', () => {
+      modDB.addMod(createMod({ name: 'MindOverMatter', type: 'BASE', value: 150 }));
+
+      const result = calculateDefence(env);
+
+      expect(result.mindOverMatterPercent).toBe(100);
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/calculator/DefenceCalculator.test.ts]: Fixed in 953ef16. Added test that verifies 100% MoM calculates EHP without division by zero.

File: src/engine/calculator/DefenceCalculator.ts [hunk 5]
```diff
@@ -0,0 +1,831 @@
...
+ * Calculate all defensive stats for a build.
+ *
+ * @param env - The calculation environment with populated ModDBs
+ * @returns Complete defence calculation results
+ */
+export function calculateDefence(env: Environment): DefenceResult {
+  const resolver = new StatResolver({
+    modDB: env.playerDB,
+    config: {
+      level: env.config.enemyLevel,
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 953ef16. Changed to use env.build.level ?? 1.

File: src/engine/calculator/DefenceCalculator.ts [hunk 6]
```diff
@@ -0,0 +1,817 @@
...
+  armour: number,
+  hitDamage: number
+): number {
+  // Start with 1 (full damage taken)
+  let multiplier = 1;
+
+  // Apply resistance
+  if (resistance > 0) {
+    multiplier *= 1 - resistance / 100;
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 6a00047. Removed the resistance > 0 check so negative resistance correctly increases damage taken (multiplier > 1).

File: src/engine/calculator/DefenceCalculator.ts [hunk 7]
```diff
@@ -0,0 +1,817 @@
...
+/**
+ * Calculate mana regeneration rate.
+ *
+ * @param resolver - StatResolver instance
+ * @param maxMana - Maximum mana
+ * @returns Mana regenerated per second
+ */
+export function calculateManaRegen(resolver: StatResolver, maxMana: number): number {
+  // Base mana regen is 1.75% per second in PoE
+  const basePercent = 1.75;
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 6a00047. Extracted to BASE_STATS.MANA_REGEN_PERCENT in shared/constants.ts.

File: src/engine/calculator/DefenceCalculator.ts [hunk 8]
```diff
@@ -0,0 +1,817 @@
...
+    lightning: Math.floor(ehpLightning),
+    chaos: Math.floor(ehpChaos),
+    elemental: Math.floor(ehpElemental),
+  };
+}
+
+/**
+ * Calculate the damage multiplier (fraction of damage taken after mitigation).
+ *
+ * @param resistance - Resistance percentage (0-90)
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/calculator/DefenceCalculator.ts]: Fixed in 6a00047. Updated JSDoc to document that resistance can be negative from curses.
